### PR TITLE
Implemented Threefold repetition check

### PR DESCRIPTION
**Pull request by joee-ca**
I just finished implementing Threefold repetition check and its relative test.

Looking a bit on the internet I found out that usually people use Zobrist Hashing to store chess positions in a dictionary and check the repetition, but I decided to use Swift Hasher for simplicity. If you later want to implement it through Zobrist you can still do it, since the hashing is done separately (if you want I can implement it when I find some time).
Other than that the function also support FivefoldRepetition. It all depends on the "times" parameter used in the function.
I tried to respect your will to delegate the check to the Board, but I'm concerned about the actual possibility of storing positions that are a result of game PGN moves into the gamePositions dictionary inside "Board".

I hope everything is fine. Tell me if there is any error 😊

### CODE REVIEW

File: Sources/ChessKit/Position.swift [hunk 1]
```diff
@@ -29,6 +29,16 @@ public struct Position: Equatable, Sendable {
...
+    var hashed: Int {
+        var hasher: Hasher = .init()
+        
+        hasher.combine(pieceSet)
+        hasher.combine(sideToMove)
+        hasher.combine(legalCastlings)
+        hasher.combine(enPassant)
+        
+        return hasher.finalize()
+    }
```
Comment (pdil) [inline Sources/ChessKit/Position.swift:41]: You can remove this property and just conform `Position` to `Hashable`, it will then have a `hashValue` property which will do this automatically

File: Sources/ChessKit/Position.swift [hunk 2]
```diff
@@ -198,7 +208,20 @@ public struct Position: Equatable, Sendable {
...
+    ///   - parameter times: the times the position occurred
+    ///   - parameter positions: all the positions occurred during the game
+    /// - Returns: a bool stating if the position already occurred N times
+    public func occurred(times: Int, in positions: [Int:Int]) -> Bool {
+        if positions[hashed] == times {
+            return true
+        }
+        
+        return false
+    }
```
Comment (pdil) [inline Sources/ChessKit/Position.swift:223]: If `Position` is `Hashable` this can be removed and you can do everything from inside `Board`

File: Sources/ChessKit/Board.swift [hunk 3]
```diff
@@ -22,6 +22,9 @@ public struct Board: Sendable {
     /// The current position represented on the board.
     public var position: Position
 
+    /// All the positions occurred during the game
+    public var gamePositions: [Int: Int]
```
Comment (pdil) [inline Sources/ChessKit/Board.swift:26]: I would rename this to `positionHashCounts` or something like that so it's more clear.

I would also make it `private` so it's not exposed to other users since it's more of an internal thing.

File: Sources/ChessKit/Board.swift [hunk 4]
```diff
@@ -198,13 +202,18 @@ public struct Board: Sendable {
...
 
         // checks & mate
         let checkState = self.checkState(for: move.piece.color)
         processedMove.checkState = checkState
-
+        
+        gamePositions.updateValue(
+            (gamePositions[position.hashed] ?? 0) + 1,
+            forKey: position.hashed
+        )
```
Comment (pdil) [inline Sources/ChessKit/Board.swift:215]: If the other changes are applied this can become:
```suggestion
        // repetition
        positionHashCounts[position.hashValue, default: 0] += 1
```

**Review by pdil — CHANGES_REQUESTED** [review]
Thanks! this looks great, and having looked at the performance tests it doesn't seem to have had a major effect but I will verify it

**Comment by joee-ca** [comment]
I did every change you requested, but:
1- I had to add conformance function hash to make `Position` class `Hashable`
2- Since `positionHashCounts` is now private, I had to add a function to retrieve that variable for the test

Also I fixed a problem related to the FEN notation. Basically you were adding the enPassant square to it every time a player moved a pawn by 2 squares, but that is not the only condition that allows enPassant. I added `sideSquares` variable to `Piece` to check if there were any pawns on the side of the the one that moved by 2 squares, and if so the enPassant would be added to the notation.

Hope everything is clear 😄

**Comment by joee-ca** [comment]
Thinking about it, the enPassant square should not be written inside the FEN notation also if it leads to a check, so I added another check to avoid it

File: Sources/ChessKit/Position.swift [hunk 5]
```diff
@@ -29,6 +29,14 @@ public struct Position: Equatable, Sendable {
...
     var enPassant: EnPassant?
 
+    /// Conforming Position class to Hashable
+    public func hash(into hasher: inout Hasher) {
+        hasher.combine(pieceSet)
+        hasher.combine(sideToMove)
+        hasher.combine(legalCastlings)
+        hasher.combine(enPassant)
+    }
+
```
Comment (pdil) [inline Sources/ChessKit/Position.swift:39]: If `Clock` is made `Hashable` this won't be necessary either and will happen automatically

**Review by pdil — CHANGES_REQUESTED** [review]
Can you revert the indentation changes in `Piece.swift` and `Board.swift` as well?

**Comment by pdil** [comment]
> Thinking about it, the enPassant square should not be written inside the FEN notation also if it leads to a check, so I added another check to avoid it

@joee-ca, in FEN the en passant is notated whenever a pawn has moved two squares, even if there is no piece to capture, see [here](https://en.wikipedia.org/wiki/Forsyth%E2%80%93Edwards_Notation#Definition):
> [En passant](https://en.wikipedia.org/wiki/En_passant) target square: This is a square over which a pawn has just passed while moving two squares; it is given in algebraic notation. If there is no en passant target square, this field uses the character "-". This is recorded regardless of whether there is a pawn in position to capture en passant.[[6]](https://en.wikipedia.org/wiki/Forsyth%E2%80%93Edwards_Notation#cite_note-6) An updated version of the spec has since made it so the target square is recorded only if a legal en passant capture is possible, but the old version of the standard is the one most commonly used.[[7]](https://en.wikipedia.org/wiki/Forsyth%E2%80%93Edwards_Notation#cite_note-7)[[8]](https://en.wikipedia.org/wiki/Forsyth%E2%80%93Edwards_Notation#cite_note-8)
In any case if the change were to be made I would prefer it to be in a separate PR so we can keep this one focused on the repetition checks.

**Comment by joee-ca** [comment]
I can revert all the changes and make the clock `Hashable` but there are 2 problems:

**Problem number 1**

If we keep:
```Swift
if abs(start.rank.value - end.rank.value) == 2 {
         position.enPassant = EnPassant(pawn: updatedPiece)
   }
```
in `Board` like you said we will have problem in checking for threefold repetition, since 2 positions will be considered different even tho the **POSSIBILITY TO EXECUTE** the enPassant is the same.

**Problem number 2**

Making the clock `Hashable` and using the default implementation of `.hashValue` modifier on `Position` will return every time a different value.

For point n.1, to implement things like you said, we should change the way FEN string is written or find some kind of workaround, because enPassant rights are written if `position.enPassant` is not nil:
```Swift
if let enPassant = position.enPassant {
            fen += "\(enPassant.captureSquare) "
        } else {
            fen += "- "
        }
```
Basically we should do something like `if lastMove == pawnMovedBy2` to write the FEN, because the ***possibility to execute enPassant is different from having targeted square inside the FEN***.

On the other hand, for what regards point n.2, I think it's better to keep things like they are right now (defining hash function), but suggestions are appreciated.

Tell me what you think so that I can start working on it 😄

**Comment by pdil** [comment]
@joee-ca 

Ah ok, regarding the `Clock` yes I guess we can't include that in the hash or every position will appear different, good point.

As for en passant, I'm having trouble understanding. We reset `enPassant` to `nil` if the player does not capture en passant, so on the next iteration of the move it will appear different which is correct because the positions are not considered repeated if the player was able to en passant on one turn and then not able to on another.

Quoting [here](https://en.wikipedia.org/wiki/Threefold_repetition#Statement_of_the_rule):
> Positions are considered the same if
(1) the same player has the move,
(2) pieces of the same kind and color occupy the same squares, and
(3) the possible moves of all the pieces are the same.
Under (3) above, positions are not considered to be the same if:
(a) **in the first position, a pawn could have been captured en passant (by the en passant rule, in the subsequent positions, the pawn cannot be captured en passant anymore)**, or
(b) either player has lost a right to castle, i.e. either king or one of the rooks has been moved, in between repetitions of the position.

**Comment by joee-ca** [comment]
Sorry, I'm bad at explaining things 😅. I will try to be as clear as possible.
EnPassant is not nil when a player moves a pawn by 2 squares. Now let's do a quick example:

**PGN Notation**

`1. e4 e5` -> classic pawn moves
`2. Nf3 Nf6` -> moving knights forward
`3. Ng1 Ng8` -> moving knights back

Now after `1 .e4 e5` and `3. Ng1 Ng8` the position will be exactly the same, but since on move 1 the pawn has moved from e7 to e5, the enPassant will not be nil, hence the hashed value will be different even tho it should not.

This is the reason why it brings problem to the threefold repetition check. We should divide the possibility to actually execute the enPassant from the need of writing the target square into the FEN notation. One solution could also be storing all the legal moves inside the Position, so that we don't pass enPassant + legalCastlings but only the legalMoves array.

I hope I explained myself well 😅

P.S: you should change the documentation of `enPassant` in `Position.swift`, since the value is not nil if a player moved a pawn by 2 squares, not if there is the possibility to execute the enPassant.

**Comment by pdil** [comment]
@joee-ca 

In `Board.move(pieceAt:to:)` we do the following:
``` swift
        if piece.kind == .pawn,
           let enPassant = position.enPassant,
           enPassant.pawn.color == piece.color.opposite,
           end == enPassant.captureSquare {
            position.remove(enPassant.pawn)
            position.move(piece, to: end)
            return process(move: Move(result: .capture(enPassant.pawn), piece: piece, start: start, end: end))
        } else {
            position.enPassant = nil     // prevent en passant on next turn
        }
```
which resets `enPassant` on the following move if it's not taken. Come to think of it we should probably call `position.enPassant = nil` in both cases of the if statement.

I agree the documentation needs to be updated to be more clear.

**Comment by joee-ca** [comment]
I agree that on next move It will be put to nil, but not for the current one.
I mean if I do 1.e4 e5 there should be no enPassant since there is no white pawn that can capture and go on e6

**Comment by pdil** [comment]
@joee-ca ok yeah now I see what you're saying! Sorry about the misunderstanding.

Yeah this is tricky because we still need `EnPassant` to be set whenever a pawn moves two squares for writing the FEN.

What we can do is add a property to `Position` that we modify in the `Board.move(pieceAt:to:)` function where you added the check as needed when a pawn moves by 2 squares. Something like `var enPassantIsPossible: Bool`. Then in the `Position.hash(into:)` function we check:
``` swift
if enPassantIsPossible {
    hasher.combine(enPassant)
}
```

We just have to be very careful and make sure `enPassantIsPossible` is set correctly each time a move is made. I'm open to other suggestions too but I think this way satisfies the repetition check and the FEN notation requirements.

**Comment by joee-ca** [comment]
@pdil 
I have two suggestions:

1- We could add a property to `Position` that we modify on `Board.move` called `targetSquare`. Basically its type is `Square?` and we can make It equal to `EnPassant(updatedPiece).captureSquare` after a pawn move by 2 squares.
After that we run other checks to see if enPassant is possible and, if it is, we put `position.enPassant = EnPassant(updatedPiece)`.
Than inside `FENParser` we can do `if let square = position.targetSquare { fen +=  "\(square) " } else { fen += "- "}`

2- We can add the property `enPassantIsPossible` to position, but then we don't use it like you said, because it would make no difference
```Swift
if enPassantIsPossible {
    hasher.combine(enPassant)
    // hashing or not hashing the enPassant would make no difference. The important thing is having it nil or not according to the position
}
```

We could do something like this in `Board.process`:
```Swift
if !enPassantIsPossible {
   positionHashCounts[position.hashValue, default: 0] += 1
}
```
Basically if the enPassant is possible, we don't put that position inside the positionHashCounts dictionary. Why? Because on next move it won't be available anymore. This means that every position with an enPassant possibility is unique and can appear maximum 1 time per game, so we don't need to count it for three/five-fold repetition.

Among these options, I think the 1st is the best, but for sure there is something better that I'm missing. I will try to find it in the meanwhile. Tell me if you had any other idea 🥲

**Comment by pdil** [comment]
Hey @joee-ca, sorry for the late response. I think I would prefer option 2 since it's simpler and has less things we need to keep track of. Let me know what you think.

**Comment by joee-ca** [comment]
Ehi @pdil , don't worry!
I've been kinda sick in these days but I managed to implement everything. I'm still not fully convinced about the solution we went for, but I agree that it's simple. I had a simpler solution in mind, that is this one:
```Swift
private func validate(moveFor piece: Piece, to square: Square) -> Bool {
        // attempt move in test set
        var testSet = set
        testSet.remove(piece)
        
        var movedPiece = piece
        movedPiece.square = square
        testSet.add(movedPiece)
        
        if let enPassant = position.enPassant {
            if enPassant.couldBeCaptured(by: piece) && enPassant.captureSquare == square {
                testSet.remove(enPassant.pawn)
                 // New lines here
                 if !position.enPassantIsPossible {
                    position.enPassantIsPossible = !isKingInCheck(piece.color, set: testSet)
                }
                // end.
            }
        }
        
        return !isKingInCheck(piece.color, set: testSet)
    }
    ```
Basically I wanted to add those new lines to avoid creating the function "validateEnPassant", but I had to mark every function as `mutating` and I don't know if you are ok with it.
The `if !position.enPassantIsPossible` is there because maybe two pawns could be able to execute enPassant, but only one actually can. This ensures that if I already checked the pawn who can execute it inside the set, when the next one comes it won't set the possibility to false.

**Square left/right**
I know you said to remove sideSquares from Piece and I agree that was shit, but Square left and right are really useful to avoid writing extra code. One thing you could notice is that they are not optional, but why? When the file number becomes 0 or 9, it will go into the default case of the enum, that basically means 1 (File A). Since these variables are only used to check the enPassant, when it falls into the default case it won't make it through the checks of `coulbBeCaptured`, so it is still safe (but if you want I can make that optional).
It's kinda like the enPassant check inside the `validate` function. If `position.enPassant != nil` you basically check if the piece conforms to the enPassant capturing pawn and if it is moving to the enPassant capture square, but you actually do the check for every piece. This is also the main reason why I'm not convinced about this solution. I don't think we should check if every peace conforms to the capturingPawn if there is an enPassant, but it is a special move and for now I have no solutions for that 😥 

If you have any questions/request just tell me. I still have to implement tests because I first wanted to know what you think about it. Let me know 😄

File: Sources/ChessKit/Special Moves/EnPassant.swift [hunk 6]
```diff
@@ -14,16 +14,16 @@ struct EnPassant: Equatable, Hashable, Sendable {
...
     /// - parameter capturingPiece: The piece that is capturing the contained pawn.
-    /// - returns: Whether `capturingPiece` can capture `pawn`.
+    /// - returns: Whether `capturingPiece` could capture `pawn`.
     ///
     /// `capturingPiece` must be an opposite color pawn that is on the
     /// same rank as the target pawn and exactly 1 file away from the
     /// target pawn for this method to return `true`, otherwise `false`
     /// is returned.
-    func canBeCaptured(by capturingPiece: Piece) -> Bool {
+    public func couldBeCaptured(by capturingPiece: Piece) -> Bool {
```
Comment (pdil) [inline Sources/ChessKit/Special Moves/EnPassant.swift:26]: Does this need to be `public`? `EnPassant` isn't public right now so I don't think this will have any effect.

File: Sources/ChessKit/Square.swift [hunk 7]
```diff
@@ -60,6 +60,16 @@ public enum Square: Int, Equatable, CaseIterable, Sendable {
...
+    
+    /// The Square at the left of the current one.
+    public var left: Square {
+        return Square(File(self.file.number - 1), self.rank)
+    }
+    
+    /// The Square at the right of the current one.
+    public var right: Square {
+        return Square(File(self.file.number + 1), self.rank)
+    }
```
Comment (pdil) [inline Sources/ChessKit/Square.swift:72]: Just some minor suggestions so that the code matches the style of the project:

```suggestion
    /// The Square at the left of the current one.
    public var left: Square {
        Square(File(file.number - 1), rank)
    }
    
    /// The Square at the right of the current one.
    public var right: Square {
        Square(File(file.number + 1), rank)
    }
```

File: Tests/ChessKitTests/BoardTests.swift [hunk 8]
```diff
@@ -152,29 +161,54 @@ class BoardTests: XCTestCase {
...
+        
+        board.move(pieceAt: .g1, to: .f3)
+        board.move(pieceAt: .g8, to: .f6)
+        
+        board.move(pieceAt: .f3, to: .g1)
+        
+        // check draw before and after the move
+        XCTAssertFalse(board.hashedPositions()[board.position.hashValue] == 3)
+        board.move(pieceAt: .f6, to: .g8)
+        XCTAssertTrue(board.hashedPositions()[board.position.hashValue] == 3)
```
Comment (pdil) [inline Tests/ChessKitTests/BoardTests.swift:201]: Instead of checking `hashedPositions`, I would use a mock `BoardDelegate`. See `testInsufficientMaterial()` and `testFiftyMoveRule()` for examples. Then `hashedPositions()` can be removed and we don't have to expose the implementation details.

File: Sources/ChessKit/Board.swift [hunk 9]
```diff
@@ -136,19 +141,26 @@ public struct Board: Sendable {
...
             }
-
             return process(move: move)
         }
     }
-
+    
+    public func hashedPositions() -> [Int:Int] {
+        return self.positionHashCounts
+    }
```
Comment (pdil) [inline Sources/ChessKit/Board.swift:163]: I don't think we need this function, see my suggestion for the tests so we don't need to expose `positionHashCounts` publicly.

**Review by pdil — CHANGES_REQUESTED** [review]
Looking good! I think this is now at a reasonable solution that works well.

The only thing I would add is can you revert the indentation changes that were made in `Board`, `Piece`, `Position`, and `BoardTests`?

I also have a few other minor changes suggested as well:

**Comment by joee-ca** [comment]
@pdil done!
Everything you asked for is there. Just tell me if I missed something 😄

**Review by pdil — APPROVED** [review]
